### ISSUE TITLE
Can't get to compile w/ focus addHook

### ISSUE THREAD

**Issue by tapickell** [keyboardio/Kaleidoscope-Focus#13]
```
WARNING: Category 'Debug' in library Kaleidoscope-Hardware-Virtual is not valid. Setting to 'Uncategorized'
sketch/Model01-Firmware.ino.cpp.o:(.data._ZZ5setupE2_c+0x2): undefined reference to `setup::__c'
collect2: error: ld returned 1 exit status
Multiple libraries were found for "HID.h"
 Used: /Users/tpickell/Library/Arduino15/packages/keyboardio/hardware/avr/1.91.0-beta/libraries/HID
 Not used: /private/var/folders/bd/cg9f_zr16574xtlh_0nmf5640000gp/T/AppTranslocation/197AE6E7-7587-4177-B214-DD1A7A9522CD/d/Arduino.app/Contents/Java/hardware/arduino/avr/libraries/HID
exit status 1
Error compiling for board Keyboardio Model 01.
```

**Comment by tapickell** [comment]
I am following the example to try to use this with my keyboard. everything else works when I comment out the line for the addHook method call in setup.
I am not experienced debugging arduino c code yet.

**Comment by algernon** [comment]
Can you share your full source, by any chance? That'd help me in reproducing the issue.

**Comment by tapickell** [comment]
```
// -*- mode: c++ -*-
// Copyright 2016 Keyboardio, inc. <jesse@keyboard.io>
// See "LICENSE" for license details

#ifndef BUILD_INFORMATION
#define BUILD_INFORMATION "locally built"
#endif

// The Kaleidoscope core
#include "Kaleidoscope.h"

#include <Kaleidoscope-TopsyTurvy.h>
#include <Kaleidoscope-Focus.h>

#include "Kaleidoscope-MouseKeys.h"
#include "Kaleidoscope-Macros.h"
#include "Kaleidoscope-LEDControl.h"
#include "Kaleidoscope-NumPad.h"
#include "LED-Off.h"
#include "Kaleidoscope-LEDEffect-BootGreeting.h"
#include "Kaleidoscope-LEDEffect-SolidColor.h"
#include "Kaleidoscope-LEDEffect-Breathe.h"
#include "Kaleidoscope-LEDEffect-Chase.h"
#include "Kaleidoscope-LEDEffect-Rainbow.h"
#include "Kaleidoscope-LED-Stalker.h"
#include "Kaleidoscope-LED-AlphaSquare.h"
#include "Kaleidoscope-Model01-TestMode.h"
#include "Kaleidoscope-HostPowerManagement.h"
#include "Kaleidoscope-MagicCombo.h"
#include "Kaleidoscope-USB-Quirks.h"

enum { MACRO_VERSION_INFO,
       MACRO_ANY,
       MACRO_FORMAT,
       MACRO_REG,
       MACRO_EXEC_READ,
       MACRO_WRITE,
       MACRO_QUIT,
       MACRO_WRITE_QUIT,
       MACRO_VSPLIT,
       MACRO_HSPLIT,
       MACRO_COLON
     };

enum { PRIMARY, NUMPAD, FUNCTION }; // layers

#define PRIMARY_KEYMAP_QWERTY
// #define PRIMARY_KEYMAP_COLEMAK
// #define PRIMARY_KEYMAP_DVORAK
// #define PRIMARY_KEYMAP_CUSTOM



/* This comment temporarily turns off astyle's indent enforcement
     so we can make the keymaps actually resemble the physical key layout better
*/
// *INDENT-OFF*

KEYMAPS(

#if defined (PRIMARY_KEYMAP_QWERTY)
  [PRIMARY] = KEYMAP_STACKED
  (___,          Key_1, Key_2, Key_3, Key_4, Key_5, Key_LEDEffectNext,
   Key_Backtick, Key_Q, Key_W, Key_E, Key_R, Key_T, Key_Tab,
   Key_PageUp,   Key_A, Key_S, Key_D, Key_F, Key_G,
   Key_PageDown, Key_Z, Key_X, Key_C, Key_V, Key_B, Key_Escape,
   Key_LeftControl, Key_Backspace, Key_LeftGui, Key_LeftShift,
   Key_Escape,

   M(MACRO_ANY),  Key_6, Key_7, Key_8,     Key_9,         Key_0,         ___,
   Key_Enter,     Key_Y, Key_U, Key_I,     Key_O,         Key_P,         Key_Equals,
   Key_H, Key_J, Key_K,     Key_L,         TOPSY(Semicolon), Key_Quote,
   ShiftToLayer(FUNCTION),  Key_N, Key_M, Key_Comma, Key_Period,    Key_Slash,     Key_Minus,
   Key_RightShift, Key_LeftAlt, Key_Spacebar, Key_RightControl,
   Key_Spacebar),

#elif defined (PRIMARY_KEYMAP_DVORAK)

  [PRIMARY] = KEYMAP_STACKED
  (___,          Key_1,         Key_2,     Key_3,      Key_4, Key_5, Key_LEDEffectNext,
   Key_Backtick, Key_Quote,     Key_Comma, Key_Period, Key_P, Key_Y, Key_Tab,
   Key_PageUp,   Key_A,         Key_O,     Key_E,      Key_U, Key_I,
   Key_PageDown, Key_Semicolon, Key_Q,     Key_J,      Key_K, Key_X, Key_Escape,
   Key_LeftControl, Key_Backspace, Key_LeftGui, Key_LeftShift,
   ShiftToLayer(FUNCTION),

   M(MACRO_ANY),   Key_6, Key_7, Key_8, Key_9, Key_0, LockLayer(NUMPAD),
   Key_Enter,      Key_F, Key_G, Key_C, Key_R, Key_L, Key_Slash,
                   Key_D, Key_H, Key_T, Key_N, Key_S, Key_Minus,
   Key_RightAlt,   Key_B, Key_M, Key_W, Key_V, Key_Z, Key_Equals,
   Key_RightShift, Key_LeftAlt, Key_Spacebar, Key_RightControl,
   ShiftToLayer(FUNCTION)),

#elif defined (PRIMARY_KEYMAP_COLEMAK)

  [PRIMARY] = KEYMAP_STACKED
  (___,          Key_1, Key_2, Key_3, Key_4, Key_5, Key_LEDEffectNext,
   Key_Backtick, Key_Q, Key_W, Key_F, Key_P, Key_G, Key_Tab,
   Key_PageUp,   Key_A, Key_R, Key_S, Key_T, Key_D,
   Key_PageDown, Key_Z, Key_X, Key_C, Key_V, Key_B, Key_Escape,
   Key_LeftControl, Key_Backspace, Key_LeftGui, Key_LeftShift,
   ShiftToLayer(FUNCTION),

   M(MACRO_ANY),  Key_6, Key_7, Key_8,     Key_9,         Key_0,         LockLayer(NUMPAD),
   Key_Enter,     Key_J, Key_L, Key_U,     Key_Y,         Key_Semicolon, Key_Equals,
                  Key_H, Key_N, Key_E,     Key_I,         Key_O,         Key_Quote,
   Key_RightAlt,  Key_K, Key_M, Key_Comma, Key_Period,    Key_Slash,     Key_Minus,
   Key_RightShift, Key_LeftAlt, Key_Spacebar, Key_RightControl,
   ShiftToLayer(FUNCTION)),

#elif defined (PRIMARY_KEYMAP_CUSTOM)
  // Edit this keymap to make a custom layout
  [PRIMARY] = KEYMAP_STACKED
  (___,          Key_1, Key_2, Key_3, Key_4, Key_5, Key_LEDEffectNext,
   Key_Backtick, Key_Q, Key_W, Key_E, Key_R, Key_T, Key_Tab,
   Key_PageUp,   Key_A, Key_S, Key_D, Key_F, Key_G,
   Key_PageDown, Key_Z, Key_X, Key_C, Key_V, Key_B, Key_Escape,
   Key_LeftControl, Key_Backspace, Key_LeftGui, Key_LeftShift,
   ShiftToLayer(FUNCTION),

   M(MACRO_ANY),  Key_6, Key_7, Key_8,     Key_9,         Key_0,         LockLayer(NUMPAD),
   Key_Enter,     Key_Y, Key_U, Key_I,     Key_O,         Key_P,         Key_Equals,
                  Key_H, Key_J, Key_K,     Key_L,         Key_Semicolon, Key_Quote,
   Key_RightAlt,  Key_N, Key_M, Key_Comma, Key_Period,    Key_Slash,     Key_Minus,
   Key_RightShift, Key_LeftAlt, Key_Spacebar, Key_RightControl,
   ShiftToLayer(FUNCTION)),

#else

#error "No default keymap defined. You should make sure that you have a line like '#define PRIMARY_KEYMAP_QWERTY' in your sketch"

#endif


  [NUMPAD] =  KEYMAP_STACKED
  (___, ___, ___, ___, ___, ___, ___,
   ___, ___, ___, ___, ___, ___, ___,
   ___, ___, ___, ___, ___, ___,
   ___, ___, ___, ___, ___, ___, ___,
   ___, ___, ___, ___,
   ___,

   M(MACRO_VERSION_INFO),  ___, Key_Keypad7, Key_Keypad8,   Key_Keypad9,        Key_KeypadSubtract, ___,
   ___,                    ___, Key_Keypad4, Key_Keypad5,   Key_Keypad6,        Key_KeypadAdd,      ___,
                           ___, Key_Keypad1, Key_Keypad2,   Key_Keypad3,        Key_Equals,         ___,
   ___,                    ___, Key_Keypad0, Key_KeypadDot, Key_KeypadMultiply, Key_KeypadDivide,   Key_Enter,
   ___, ___, ___, ___,
   ___),

  [FUNCTION] =  KEYMAP_STACKED
  (___,      M(MACRO_EXEC_READ),           Key_F2,      Key_F3,     Key_F4,        Key_F5,           ___,
   Key_Tab,  M(MACRO_QUIT),    M(MACRO_WRITE), ___,     M(MACRO_REG), Key_mouseWarpEnd, Key_mouseWarpNE,
   Key_Home, Key_mouseL,       Key_mouseDn, Key_mouseR, M(MACRO_FORMAT), Key_mouseWarpNW,
   Key_End,  Key_PrintScreen,  Key_Insert,  ___,        Key_mouseBtnM, Key_mouseWarpSW,  M(MACRO_WRITE_QUIT),
   ___, Key_Delete, ___, Key_CapsLock,
   ___,

   Key_Pipe,                   Key_F6,                 Key_F7,                   Key_F8,                   Key_F9,          Key_F10,          LockLayer(NUMPAD),
   Consumer_PlaySlashPause,    Consumer_ScanNextTrack, Key_LeftCurlyBracket,     Key_RightCurlyBracket,    Key_LeftBracket, Key_RightBracket, Key_F11,
                               Key_LeftArrow,          Key_DownArrow,            Key_UpArrow,              Key_RightArrow,  Key_Semicolon,    Key_F12,
   Key_Backslash,              ___,                    Consumer_Mute,            Consumer_VolumeDecrement, Consumer_VolumeIncrement,          M(MACRO_VSPLIT),    M(MACRO_HSPLIT),
   ___, ___, Key_Enter, ___,
   ___)
  ) // KEYMAPS(

/* Re-enable astyle's indent enforcement */
// *INDENT-ON*

/** versionInfoMacro handles the 'firmware version info' macro
    When a key bound to the macro is pressed, this macro
    prints out the firmware build information as virtual keystrokes
*/

static void versionInfoMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR("Keyboardio Model 01 - Kaleidoscope "));
    Macros.type(PSTR(BUILD_INFORMATION));
  }
}

static void enterKey(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    kaleidoscope::hid::pressKey(Key_Enter);
  }
}

static void mixFormatMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR("mix format"));
  }
}

static void vimRegMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":reg"));
  }
}

static void vimExecReadMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":r!"));
  }
}

static void vimVsplitMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":vsplit"));
  }
}

static void vimHsplitMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":split"));
  }
}

static void vimWriteMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":w"));
  }
}

static void vimQuitMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":q"));
  }
}

static void vimWriteQuitMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":wq"));
  }
}

static void colonMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR(":"));
  }
}

static void anyKeyMacro(uint8_t keyState) {
  if (keyToggledOn(keyState)) {
    Macros.type(PSTR("|>"));
  }
}

const macro_t *macroAction(uint8_t macroIndex, uint8_t keyState) {
  switch (macroIndex) {

    case MACRO_VERSION_INFO:
      versionInfoMacro(keyState);
      break;

    case MACRO_ANY:
      anyKeyMacro(keyState);
      break;

    case MACRO_FORMAT:
      mixFormatMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_REG:
      vimRegMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_EXEC_READ:
      vimExecReadMacro(keyState);
      break;

    case MACRO_WRITE:
      vimWriteMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_QUIT:
      vimQuitMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_WRITE_QUIT:
      vimWriteQuitMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_VSPLIT:
      vimVsplitMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_HSPLIT:
      vimHsplitMacro(keyState);
      enterKey(keyState);
      break;

    case MACRO_COLON:
      colonMacro(keyState);
      break;
  }
  return MACRO_NONE;
}

static kaleidoscope::LEDSolidColor solidRed(160, 0, 0);
static kaleidoscope::LEDSolidColor solidOrange(140, 70, 0);
static kaleidoscope::LEDSolidColor solidYellow(130, 100, 0);
static kaleidoscope::LEDSolidColor solidGreen(0, 160, 0);
static kaleidoscope::LEDSolidColor solidBlue(0, 70, 130);
static kaleidoscope::LEDSolidColor solidIndigo(0, 0, 170);
static kaleidoscope::LEDSolidColor solidViolet(130, 0, 120);

void toggleLedsOnSuspendResume(kaleidoscope::HostPowerManagement::Event event) {
  switch (event) {
    case kaleidoscope::HostPowerManagement::Suspend:
      LEDControl.paused = true;
      LEDControl.set_all_leds_to({0, 0, 0});
      LEDControl.syncLeds();
      break;
    case kaleidoscope::HostPowerManagement::Resume:
      LEDControl.paused = false;
      LEDControl.refreshAll();
      break;
    case kaleidoscope::HostPowerManagement::Sleep:
      break;
  }
}

/** hostPowerManagementEventHandler dispatches power management events (suspend,
   resume, and sleep) to other functions that perform action based on these
   events.
*/
void hostPowerManagementEventHandler(kaleidoscope::HostPowerManagement::Event event) {
  toggleLedsOnSuspendResume(event);
}

/** This 'enum' is a list of all the magic combos used by the Model 01's
   firmware The names aren't particularly important. What is important is that
   each is unique.

   These are the names of your magic combos. They will be used by the
   `USE_MAGIC_COMBOS` call below.
*/
enum {
  // Toggle between Boot (6-key rollover; for BIOSes and early boot) and NKRO
  // mode.
  COMBO_TOGGLE_NKRO_MODE
};

/** A tiny wrapper, to be used by MagicCombo.
   This simply toggles the keyboard protocol via USBQuirks, and wraps it within
   a function with an unused argument, to match what MagicCombo expects.
*/
static void toggleKeyboardProtocol(uint8_t combo_index) {
  USBQuirks.toggleKeyboardProtocol();
}

/** Magic combo list, a list of key combo and action pairs the firmware should
   recognise.
*/
USE_MAGIC_COMBOS({.action = toggleKeyboardProtocol,
                  // Left Fn + Esc + Shift
                  .keys = { R3C6, R2C6, R3C7 }
                 });

// First, tell Kaleidoscope which plugins you want to use.
// The order can be important. For example, LED effects are
// added in the order they're listed here.
KALEIDOSCOPE_INIT_PLUGINS(
  BootGreetingEffect,
  TestMode,
  LEDControl,
  LEDOff,
  LEDRainbowEffect,
  LEDRainbowWaveEffect,
  LEDChaseEffect,
  // These static effects turn your keyboard's LEDs a variety of colors
  solidRed, solidOrange, solidYellow, solidGreen, solidBlue, solidIndigo, solidViolet,
  LEDBreatheEffect,
  AlphaSquareEffect,
  StalkerEffect,
  NumPad,
  Macros,
  MouseKeys,
  HostPowerManagement,
  MagicCombo,
  USBQuirks,
  Focus,
  TopsyTurvy
);

void setup() {
  Kaleidoscope.setup();
  NumPad.numPadLayer = NUMPAD;
  AlphaSquare.color = CRGB(255, 0, 0);
  LEDRainbowEffect.brightness(150);
  LEDRainbowWaveEffect.brightness(150);
  StalkerEffect.variant = STALKER(Rainbow);
  LEDOff.activate();

  Focus.addHook(FOCUS_HOOK_HELP);
  // Focus.addHook(FOCUS_HOOK_VERSION);
  // Focus.addHook(FOCUS_HOOK_LEDCONTROL);
}

void loop() {
  Kaleidoscope.loop();
}
```

**Comment by ryansch** [comment]
I'm running into the same problem on the same version of the boards file (1.91.0-beta).

**Comment by obra** [comment]
I think I've managed to figure out the issue. I suspect that when you
installed 1.91, an error flashed by telling you that Windows couldn't find
the program 'ln'

1.92 should fix this, which -should- be later today.
ᐧ

**Comment by obra** [comment]
1.92 is now out. I believe it fixes this.
ᐧ

**Comment by algernon** [comment]
As we've moved to `FocusSerial`, which doesn't have the `addHook` method either, I believe this can be closed, even if it wasn't fixed with 1.92.
